File: 2024/01/b.mjs
import fs from "node:fs";

const compareNumbers = (a, b) => a - b;

function main(inputFile = "in") {
  const locations = { left: [], right: [] };
  fs.readFileSync(inputFile)
    .toString()
    .trim()
    .split("\n")
    .forEach((line) => {
      const [l, r] = line.split("   ").map(Number);
      locations.left.push(l);
      locations.right.push(r);
    });
  locations.left.sort(compareNumbers);
  locations.right.sort(compareNumbers);
  const rightMap = {};
  for (const n of locations.right) {
    rightMap[n] = (rightMap[n] ?? 0) + 1;
  }
  let result = 0;
  for (const n of locations.left) {
    result += n * (rightMap[n] ?? 0);
  }
  console.log(result);
}

main();

File: 2024/01/src/b.rs
use std::collections::HashMap;
use std::fs;

pub fn main() -> i32 {
  let mut locations_left: Vec<i32> = Vec::new();
  let mut locations_right: Vec<i32> = Vec::new();
  fs::read_to_string("input/in").unwrap().trim().lines()
    .for_each(|line| {
      let mut split = line.split("   ");
      locations_left.push(split.next().unwrap().parse::<i32>().unwrap());
      locations_right.push(split.next().unwrap().parse::<i32>().unwrap());
    });
  locations_left.sort_unstable();
  locations_right.sort_unstable();
  let mut right_map: HashMap<i32, i32> = HashMap::new();
  for n in &locations_right {
    right_map.entry(*n).and_modify(|e| *e += 1).or_insert(1);
  }
  let mut result = 0;
  for n in &locations_left {
    result += n * right_map.get(n).unwrap_or(&0);
  }
  result
}
